aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -0,0 +1,877 @@
+      NULL,
+      NULL,
+      NULL,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_32I,
+      CUSPARSE_INDEX_BASE_ZERO,
+      cuda_data_type));
+  //--------------------------------------------------------------------------
+  // SpGEMM Computation
+  cusparseSpGEMMDescr_t spgemmDesc;
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_createDescr(&spgemmDesc));
+
+  // ask bufferSize1 bytes for external memory
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_workEstimation(
+      handle,
+      opA,
+      opB,
+      &alpha,
+      matA,
+      matB,
+      &beta,
+      matC,
+      computeType,
+      CUSPARSE_SPGEMM_DEFAULT,
+      spgemmDesc,
+      &bufferSize1,
+      NULL));
+  
+  auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
+
+  at::DataPtr dataPtr1 = allocator.allocate(bufferSize1);
+  dBuffer1 = dataPtr1.get();
+  // inspect the matrices A and B to understand the memory requiremnent for
+  // the next step
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_workEstimation(
+      handle,
+      opA,
+      opB,
+      &alpha,
+      matA,
+      matB,
+      &beta,
+      matC,
+      computeType,
+      CUSPARSE_SPGEMM_DEFAULT,
+      spgemmDesc,
+      &bufferSize1,
+      dBuffer1));
+
+  // ask bufferSize2 bytes for external memory
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_compute(
+      handle,
+      opA,
+      opB,
+      &alpha,
+      matA,
+      matB,
+      &beta,
+      matC,
+      computeType,
+      CUSPARSE_SPGEMM_DEFAULT,
+      spgemmDesc,
+      &bufferSize2,
+      NULL));
+
+  at::DataPtr dataPtr2 = allocator.allocate(bufferSize2);
+  dBuffer2 = dataPtr2.get();
+
+  // compute the intermediate product of A * B
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_compute(
+      handle,
+      opA,
+      opB,
+      &alpha,
+      matA,
+      matB,
+      &beta,
+      matC,
+      computeType,
+      CUSPARSE_SPGEMM_DEFAULT,
+      spgemmDesc,
+      &bufferSize2,
+      dBuffer2));
+  // get matrix C non-zero entries C_num_nnz1
+  int64_t C_num_rows1, C_num_cols1, C_num_nnz1;
+  TORCH_CUDASPARSE_CHECK(
+      cusparseSpMatGetSize(matC, &C_num_rows1, &C_num_cols1, &C_num_nnz1));
+  // allocate matrix C
+  // allocate C offsets
+  out.nnz_ = C_num_nnz1;
+
+  out.csr_indices_ = at::empty({out.nnz_}, output_indices.options().dtype(kInt));
+  out.csr_values_ = at::empty({out.nnz_}, output_values.options());
+  dC_columns = out.csr_indices_.data_ptr<int>();
+  dC_values = out.csr_values_.data_ptr<scalar_t>();
+  
+  // update matC with the new pointers
+  TORCH_CUDASPARSE_CHECK(
+      cusparseCsrSetPointers(matC, dC_csrOffsets, dC_columns, dC_values));
+
+  // copy the final products to the matrix C
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_copy(
+      handle,
+      opA,
+      opB,
+      &alpha,
+      matA,
+      matB,
+      &beta,
+      matC,
+      computeType,
+      CUSPARSE_SPGEMM_DEFAULT,
+      spgemmDesc));
+
+  // destroy matrix/vector descriptors
+  TORCH_CUDASPARSE_CHECK(cusparseSpGEMM_destroyDescr(spgemmDesc));
```

Comment:
Since you are not doing RAII-initialization of descriptors, these lines may never be reached if an error happens, and the descriptors will be leaked. Please take a look at how cudnn-related descriptors are handled to avoid that.